packages/solana/src/client.ts

```diff
@@ -394,38 +406,98 @@ export class Web3Modal extends Web3ModalScaffold {
         provider.chains.find(chain => chain.chainId === SolStoreUtil.state.currentChain?.chainId) ||
         provider.chains[0]
 
-      if (connectionChain) {
-        await this.switchNetwork(
-          SolHelpersUtil.getChainFromCaip(this.chains, `solana:${connectionChain.chainId}`)
-        )
-      } else {
+      if (!connectionChain) {
         provider.disconnect()
         throw new Error('The wallet does not support any of the required chains')
       }
 
-      SolStoreUtil.setIsConnected(true)
+      SolStoreUtil.setAddress(address)
+      await this.switchNetwork(
+        SolHelpersUtil.getChainFromCaip(this.chains, `solana:${connectionChain.chainId}`)
+      )
       SolStoreUtil.setProvider(provider)
       this.provider = provider
-      this.setAddress(address)
 
       window?.localStorage.setItem(SolConstantsUtil.WALLET_ID, provider.name)
 
       await this.setApprovedCaipNetworksData()
 
       this.watchProvider(provider)
+      SolStoreUtil.setIsConnected(true)
     } finally {
       this.setLoading(false)
     }
   }
 
   private watchProvider(provider: Provider) {
+    /*
+     * The auth RPC request handlers should be moved to the primary scaffold (Web3ModalScaffold).
+     * They are replicated in wagmi and ethers clients and the behavior should be kept the same
+     * between any client.
+     */
```

Comment:
Agreed. @enesozturk @svenvoskamp we should apply this in MC directly I'd think (`AppKit` class)